File: repositories/job_repository.go
package repositories

import (
	"context"
	"errors"
	"time"

	"doctor-service/entities"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type JobRepository interface {
	Create(ctx context.Context, job *entities.VerificationJob) error
	FetchNextQueuedJob(ctx context.Context) (*entities.VerificationJob, error)
	UpdateStatus(ctx context.Context, jobID uuid.UUID, status entities.JobStatus, lastError *string) error
	MarkRunning(ctx context.Context, jobID uuid.UUID) error
	MarkCompleted(ctx context.Context, jobID uuid.UUID) error
	MarkFailed(ctx context.Context, jobID uuid.UUID, errStr string) error
	FindByDoctorID(ctx context.Context, doctorID uuid.UUID) ([]entities.VerificationJob, error)
	FindByID(ctx context.Context, jobID uuid.UUID) (*entities.VerificationJob, error)
}

type GormJobRepository struct {
	db *gorm.DB
}

func NewJobRepository(db *gorm.DB) JobRepository {
	return &GormJobRepository{db: db}
}

func (r *GormJobRepository) Create(ctx context.Context, job *entities.VerificationJob) error {
	return r.db.WithContext(ctx).Create(job).Error
}

func (r *GormJobRepository) FetchNextQueuedJob(ctx context.Context) (*entities.VerificationJob, error) {
	var job entities.VerificationJob
	err := r.db.WithContext(ctx).
		Where("status = ?", entities.JobStatusQueued).
		Order("priority DESC, created_at ASC").
		First(&job).Error

	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &job, err
}

func (r *GormJobRepository) UpdateStatus(ctx context.Context, jobID uuid.UUID, status entities.JobStatus, lastError *string) error {
	updates := map[string]interface{}{
		"status": status,
	}
	if lastError != nil {
		updates["last_error"] = *lastError
	}
	return r.db.WithContext(ctx).Model(&entities.VerificationJob{}).Where("job_id = ?", jobID).Updates(updates).Error
}

func (r *GormJobRepository) MarkRunning(ctx context.Context, jobID uuid.UUID) error {
	now := time.Now()
	return r.db.WithContext(ctx).Model(&entities.VerificationJob{}).
		Where("job_id = ?", jobID).
		Updates(map[string]interface{}{
			"status":     entities.JobStatusRunning,
			"started_at": &now,
		}).Error
}

func (r *GormJobRepository) MarkCompleted(ctx context.Context, jobID uuid.UUID) error {
	now := time.Now()
	return r.db.WithContext(ctx).Model(&entities.VerificationJob{}).
		Where("job_id = ?", jobID).
		Updates(map[string]interface{}{
			"status":       entities.JobStatusSuccess,
			"completed_at": &now,
		}).Error
}

func (r *GormJobRepository) MarkFailed(ctx context.Context, jobID uuid.UUID, errStr string) error {
	now := time.Now()
	return r.db.WithContext(ctx).Model(&entities.VerificationJob{}).
		Where("job_id = ?", jobID).
		Updates(map[string]interface{}{
			"status":       entities.JobStatusFailed,
			"last_error":   errStr,
			"completed_at": &now,
		}).Error
}

func (r *GormJobRepository) FindByDoctorID(ctx context.Context, doctorID uuid.UUID) ([]entities.VerificationJob, error) {
	var jobs []entities.VerificationJob
	err := r.db.WithContext(ctx).Where("doctor_id = ?", doctorID).Order("created_at DESC").Find(&jobs).Error
	return jobs, err
}

func (r *GormJobRepository) FindByID(ctx context.Context, jobID uuid.UUID) (*entities.VerificationJob, error) {
	var job entities.VerificationJob
	err := r.db.WithContext(ctx).Where("job_id = ?", jobID).First(&job).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &job, err
}

File: python_backend/repositories/job_repository.py
import datetime
from typing import List, Optional
from uuid import UUID
from sqlalchemy.orm import Session
from entities.job import VerificationJob

class JobRepository:
    def __init__(self, db: Session):
        self.db = db

    def Create(self, job: VerificationJob) -> None:
        self.db.add(job)
        self.db.commit()
        self.db.refresh(job)

    def FetchNextQueuedJob(self) -> Optional[VerificationJob]:
        # Assuming entities.JobStatusQueued or 'QUEUED' string is used, adjust as necessary
        return self.db.query(VerificationJob).filter(
            VerificationJob.status == "QUEUED"  # Replace with appropriate constant if available
        ).order_by(VerificationJob.priority.desc(), VerificationJob.created_at.asc()).first()

    def UpdateStatus(self, job_id: UUID, status: str, last_error: Optional[str] = None) -> None:
        updates = {"status": status}
        if last_error is not None:
            updates["last_error"] = last_error
        self.db.query(VerificationJob).filter(VerificationJob.job_id == job_id).update(updates)
        self.db.commit()

    def MarkRunning(self, job_id: UUID) -> None:
        now = datetime.datetime.now()
        self.db.query(VerificationJob).filter(VerificationJob.job_id == job_id).update({
            "status": "RUNNING",
            "started_at": now
        })
        self.db.commit()

    def MarkCompleted(self, job_id: UUID) -> None:
        now = datetime.datetime.now()
        self.db.query(VerificationJob).filter(VerificationJob.job_id == job_id).update({
            "status": "SUCCESS",
            "completed_at": now
        })
        self.db.commit()

    def MarkFailed(self, job_id: UUID, err_str: str) -> None:
        now = datetime.datetime.now()
        self.db.query(VerificationJob).filter(VerificationJob.job_id == job_id).update({
            "status": "FAILED",
            "last_error": err_str,
            "completed_at": now
        })
        self.db.commit()

    def FindByDoctorID(self, doctor_id: UUID) -> List[VerificationJob]:
        return self.db.query(VerificationJob).filter(VerificationJob.doctor_id == doctor_id).order_by(VerificationJob.created_at.desc()).all()

    def FindByID(self, job_id: UUID) -> Optional[VerificationJob]:
        return self.db.query(VerificationJob).filter(VerificationJob.job_id == job_id).first()
